flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendLocalHandle.java

```diff
@@ -0,0 +1,116 @@
+
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.KeyedStateHandle;
+import org.apache.flink.runtime.state.SharedStateRegistry;
+import org.apache.flink.runtime.state.StateHandleID;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+public class ChangelogStateBackendLocalHandle implements ChangelogStateBackendHandle {
+    private static final long serialVersionUID = 1L;
+    private static final Logger LOG =
+            LoggerFactory.getLogger(ChangelogStateBackendLocalHandle.class);
+    private final List<KeyedStateHandle> localMaterialized;
+    private final List<ChangelogStateHandle> localNonMaterialized;
+    private final ChangelogStateBackendHandleImpl remoteHandle;
+
+    public ChangelogStateBackendLocalHandle(
+            List<KeyedStateHandle> localMaterialized,
+            List<ChangelogStateHandle> localNonMaterialized,
+            ChangelogStateBackendHandleImpl remoteHandle) {
+        this.localMaterialized = localMaterialized;
+        this.localNonMaterialized = localNonMaterialized;
+        this.remoteHandle = remoteHandle;
+    }
+
+    @Override
+    public List<KeyedStateHandle> getMaterializedStateHandles() {
+        return localMaterialized;
+    }
+
+    @Override
+    public List<ChangelogStateHandle> getNonMaterializedStateHandles() {
+        return localNonMaterialized;
+    }
+
+    @Override
+    public long getMaterializationID() {
+        return remoteHandle.getMaterializationID();
+    }
+
+    @Override
+    public ChangelogStateBackendHandle rebound(long checkpointId) {
+        return remoteHandle.rebound(checkpointId);
+    }
+
+    public List<KeyedStateHandle> getRemoteMaterializedStateHandles() {
+        return remoteHandle.getMaterializedStateHandles();
+    }
+
+    public List<ChangelogStateHandle> getRemoteNonMaterializedStateHandles() {
+        return remoteHandle.getNonMaterializedStateHandles();
+    }
+
+    @Override
+    public long getCheckpointId() {
+        return remoteHandle.getCheckpointId();
+    }
+
+    @Override
+    public void registerSharedStates(SharedStateRegistry stateRegistry, long checkpointID) {
+        remoteHandle.registerSharedStates(stateRegistry, checkpointID);
+    }
+
+    @Override
+    public long getCheckpointedSize() {
+        return remoteHandle.getCheckpointedSize();
+    }
+
+    @Override
+    public KeyGroupRange getKeyGroupRange() {
+        return remoteHandle.getKeyGroupRange();
+    }
+
+    @Nullable
+    @Override
+    public KeyedStateHandle getIntersection(KeyGroupRange keyGroupRange) {
+        throw new UnsupportedOperationException(
+                "This is a local state handle for the TM side only.");
+    }
+
+    @Override
+    public StateHandleID getStateHandleId() {
+        return remoteHandle.getStateHandleId();
+    }
+
+    @Override
+    public void discardState() throws Exception {}
```

Comment:
:+1: IIUC, the discard logic is correct:
1. `(Changelog)TaskLocalStateStore` calls this no-op method
1. `LocalStateRegistry` calls `discardState()` on "low-level" handle (`FileStateHandle`) - on checkpoint confirmation
1. `TaskChangelogRegistry` - calls `discardState()` on "low-level" handle (`FileStateHandle`) - when upload is no more needed
1. As opposed to "normal" local state, local changelog state is *not* discarded on recovery (because of 1); but it will be discarded on the 1st checkpoint confirmation

Can you confirm that?

Although I found it tricky to understand, so it makes sense to document it somewhere, WDYT?

(and also to cover with tests).

Reply:
Probably this case is not handled: the last checkpoint(s) aborted and then job terminated?

Reply:
> 1. (Changelog)TaskLocalStateStore calls this no-op method

Yes, `TaskLocalStateStore` call no-op method, but `TaskLocalStateStore` would delete the whole chk folder(`changelog_chk_` and `chk_`) when confirm checkpoint. 

> 4. As opposed to "normal" local state, local changelog state is not discarded on recovery (because of 1); but it will be discarded on the 1st checkpoint confirmation

Like "normal" local state, local changelog state would **be discarded** on recovery, it would be delete when `deleteDirectory()`, because of `recursive=true`.

```java
protected void deleteDirectory(File directory) throws IOException {
        Path path = new Path(directory.toURI());
        FileSystem fileSystem = path.getFileSystem();
        if (fileSystem.exists(path)) {
            fileSystem.delete(path, true);
        }
    }
```

The total discard logic of local dstl files is:
1. Register files to `TaskChangelogRegistry#startTracking()` on `StateChangeUploader#upload()`.
2.  Pass control of the file to `LocalStateRegistry#register()` **when** `confirm()` , files of the previous checkpoint will be unregistered/deleted by `LocalStateRegistry#unRegister()`  at the same time ; for remote file, we pass control of the file to JM.
3. When `materialized()` or `checkpointSubsumed()` is called, `TaskChangelogRegistry#notUse()` is responsible for deleting local files. BTW, it is possible that the file has already been truncated by `LocalStateRegistry`.

Reply:
> Yes, TaskLocalStateStore call no-op method, but TaskLocalStateStore would delete the whole chk folder(changelog_chk_ and chk_) when confirm checkpoint.

I was assuming that deleting the whole directory still does not guarantee deleting the changelog (because it can reside outside); that's why there is an explicit call `taskStateSnapshot.discardState();` in "normal" local state store.
But either way, recursive deletion or the next checkpoint should discard it.

> The total discard logic of local dstl files is

Thanks for the explanation. 
Could you add this comments ("total discard logic") somewhere int the code?

And I still don't see how this case is handled:
> the last checkpoint(s) aborted and then job terminated?

Reply:
> I was assuming that deleting the whole directory still does not guarantee deleting the changelog (because it can reside outside); that's why there is an explicit call taskStateSnapshot.discardState(); in "normal" local state store.
But either way, recursive deletion or the next checkpoint should discard it.

Yes, deleting the whole directory still does not guarantee deleting the changelog (because it can reside outside). `ChangelogStateHandleStreamImpl` are placed in `/dstl`. 
For **Local** `ChangelogStateHandleStreamImpl`, they are placed in <LOCAL_DIR>/taskowned, and they would be discarded by  `TaskChangelogRegistry` and  `LocalStateRegistry`. 
So,  *recursive deletion `changelog_chk_` directory* + *`LocalStateRegistry`* + *`TaskChangelogRegistry`* are used to guarantee non-materialized part's deletion.

> Could you add this comments ("total discard logic") somewhere in the code?

Sure, I will add this comment in `ChangelogTaskLocalStateStore`.

> Probably this case is not handled: the last checkpoint(s) aborted and then job terminated?

You are right, "the last checkpoint(s) aborted and then job terminated"  can't be handled in this code. 
Regardless of the deletion of local dstl, do you know how **remote dstl** handles this case?  if one `StateChangeSet` is uploaded to `/dstl`,  will it be removed on abort?